notebooks/ch11/gradient_stochastic.md

```diff
@@ -0,0 +1,331 @@
+import pandas as pd
+import seaborn as sns
+%matplotlib inline
+import ipywidgets as widgets
+from ipywidgets import interact, interactive, fixed, interact_manual
+import nbinteract as nbi
+
+sns.set()
+sns.set_context('talk')
+np.set_printoptions(threshold=20, precision=2, suppress=True)
+np.random.seed(42)
+pd.options.display.max_rows = 7
+pd.options.display.max_columns = 8
+pd.set_option('precision', 2)
+```
+
+
+```python
+# HIDDEN
+tips = sns.load_dataset('tips')
+tips['pcttip'] = tips['tip'] / tips['total_bill'] * 100
+```
+
+
+```python
+# HIDDEN
+def mse_loss(theta, y_vals):
+    return np.mean((y_vals - theta) ** 2)
+
+def grad_mse_loss(theta, y_vals):
+    return -2 * np.mean(y_vals - theta)
+```
+
+## A Brief Review
+
+We initially modeled the tips dataset by heuristically calculating the $\hat\theta$ that minimized the MSE loss function, $L(\hat\theta, y) = \frac{1}{n} \sum_{i=1}^{n}(y_i - \hat\theta)^2$. This involved taking the derivative of the MSE, setting it to zero, and solving for $\hat\theta$. We found that the MSE-minimizing $\hat\theta$ value was simply the mean of the $y$ values in our dataset.
+
+But for more complicated models and loss functions, there may not be simple algebraic expressions that yield the loss-minimizing $\hat\theta$ values. Our solution was a new method called gradient descent, which iteratively updates $\hat\theta$ until it doesn't change between iterations. To complete an iteration of gradient descent, we calculate the following:
+
+$$
+\hat\theta_{t+1} = \hat\theta_t - \alpha \cdot \nabla_{\hat\theta} L(\hat\theta, y)
+$$
+
+In this equation:
+- $\hat\theta_{t}$ is our current estimate of $\theta$ at the $t$th iteration
+- $\alpha$ is the learning rate
+- $\nabla_{\hat\theta} L(\hat\theta, y)$ is the gradient of the loss function
+- We compute the next estimate $\hat\theta_{t+1}$ by subtracting the product of $\alpha$ and $\nabla_{\hat\theta} L(\hat\theta, y)$ computed at $\hat\theta_{t}$
+
+## Limitations of Gradient Descent
+
+Often, we calculate $\nabla_{\hat\theta}L(\hat\theta, y)$ using the average gradient of the loss function $l(\hat\theta, y_i)$ across the entire dataset as a single batch. For this reason, this gradient update rule is often referred to as **batch gradient descent**.
+
+For example, the gradient of the MSE loss first requires us to find the gradient of the squared loss, $\nabla_{\hat\theta} l(\hat\theta, y_i) = -2 (y_i - \hat\theta)$, for each of the $n$ observations in our dataset. We then compute the average of these results. This is shown below:
+
+$$
+\begin{align}
+\nabla_{\hat\theta} L(\hat\theta, y) &= -\frac{2}{n} \sum_{i=1}^{n}(y_i - \hat\theta) \\
+&= \frac{1}{n} \sum_{i=1}^{n}-2(y_i - \hat\theta) \\
+&= \frac{1}{n} \sum_{i=1}^{n} \nabla_{\hat\theta} l(\hat\theta, y_i)
+\end{align}
+$$
+
+
+
+This process is then repeated at each iteration $t$ of the gradient descent algorithm. For large $n$, this can become a computationally expensive problem to solve.
```

Comment:
"This process" is vague. How about:

> Thus, for each iteration of batch gradient descent, we must compute the loss on every point in the dataset. Since datasets can contain many points, batch gradient descent can take a prohibitive amount of time to converge.